src/Client.re

```diff
@@ -14,10 +17,15 @@ let start =
       (),
     ) => {
   let protocolClient: ref(option(Protocol.t)) = ref(None);
+  let lastRequestId = ref(0);
```

Comment:
I agree it's not ideal, I'm open to ideas on how to improve it.

The `ref(option(Protocol.t))` is something I hit when I'm initializing an object that takes a `dispatch` or some function, that _depends on the output_ of the function. In this case - we create the `Protocol` object, but the `dispatch` we pass to `Protocol.create` depends on the object (ie, we handle the `Ready`/`Initialized` messages by sending messages back). Maybe you have a better idea in mind? I think that's the root issue - the fact that our `~dispatch` we pass to `Protocol.start` depends on the `Protocol.t`.

Reply:
Hmm, yeah I'm not sure. The more I look at this code the more confused I get. It's extremely convoluted. And although much of that is inherent to the problem, the lack of structure makes it much more difficult to follow. I think generally it would be a good idea to break it up and encapsulate the code, and the mutation in particular, as much as possible. Try to reduce the amount of code that can access the `ref`s. Instead of passing the `ref` itself into some function to be incremented, pass a function that only increment and return the next id. Break the code up into smaller functions, and either move it to the top level, or at least before the `ref`s are defined. Pass dependencies in as arguments to make them explicit. Avoid the convenience of closures that implicitly capture variables if they're more than a few lines long and/or in defined in a scope that has has mutable or many irrelevant variables. In general, reduce the number of variables that has to be kept track of to follow the code, and the span they have to be tracked across.